Added AVX512 support for space_l2 and space_ip.

slice4e opened the pull request:
Add AVX512 support to L2 and IP algorithms. 

Testing with Ann-Benchmarks shows measurable improvement on c5.4xlarge and m6i.4xlarge instances. 

c5.4xlarge results
fashion-mnist-784-euclidean
![fashion-mnist-784-euclidean_10_euclidean](https://user-images.githubusercontent.com/8259383/131779977-dbc269e2-283e-4430-8b89-b4bfaccb6c17.png)
mnist-784-euclidean
![mnist-784-euclidean_10_euclidean](https://user-images.githubusercontent.com/8259383/131779978-1684e5b6-6a3d-46d5-b6de-411fc69ba3d8.png)
nytimes-256-angular
![nytimes-256-angular_10_angular](https://user-images.githubusercontent.com/8259383/131779980-ee46d04b-2937-4c68-800d-937f80a67b9f.png)
sift-128-euclidean
![sift-128-euclidean_10_euclidean](https://user-images.githubusercontent.com/8259383/131779981-ea242b87-028a-47cb-ba25-d816277da7ab.png)


m6i.4xlarge results: 
fashion-mnist-784-euclidean
![fashion-mnist-784-euclidean_10_euclidean](https://user-images.githubusercontent.com/8259383/131780105-ee474a42-a8ac-481a-a022-a022df062f82.png)
mnist-784-euclidean
![mnist-784-euclidean_10_euclidean](https://user-images.githubusercontent.com/8259383/131780109-429a0101-9798-4ead-959e-d97eb65ace34.png)
nytimes-256-angular
![nytimes-256-angular_10_angular](https://user-images.githubusercontent.com/8259383/131780111-d1f4c266-e2c0-49cb-ac8f-f7767cde2f14.png)
sift-128-euclidean
![sift-128-euclidean_10_euclidean](https://user-images.githubusercontent.com/8259383/131780113-ac3772c1-b868-4d72-9c59-37e152769a81.png)

yurymalkov commented on hnswlib/hnswlib.h:
```diff
@@ -4,6 +4,8 @@
 #define USE_SSE
 #ifdef __AVX__
 #define USE_AVX
+#ifdef __AVX512__
```
Hm. It looks like this `#ifdef` does not have a corresponding `#endif` and this seems to trigger the CI build fail for some systems.

slice4e replied:
Yes, I did test with multiple concurrent processes running. We plugged the functions:  InnerProductSIMD16Ext and  L2SqrSIMD16Ext into a different application. The new application was implementing a brute-force vector similarity search.
The test was performed on c5.4xlarge on AWS. 

We can see that AVX512 is consistently better (lower execution time) than AVX2. We can also see that after about 8 parralel instances, the latency shoots up dramatically. We believe that this is due to memory latency. The brute-force algorithm is a memory scan and thus very memory bandwith intensive. After about 8 concurrent instances, I expect that we are saturating the memory bandwidth of the CPU. 

![IP_brute_force](https://user-images.githubusercontent.com/8259383/132893128-e55c842a-51b1-4921-a8ea-1d52ce3e0110.png)
![L2_brute_force](https://user-images.githubusercontent.com/8259383/132893131-d38f169d-5ca6-4910-a2ad-9ce381f88d15.png)
 
It is true that frequency goes down with AVX512 as compared with AVX2. Same is also true for AVX2 vs AVX. The number of cores executing AVX/AVX2/AVX512 code, determines the frequency. The 3 figures below, show the frequency tables depending on how many cores are active. As an example, I have circled a single point - assuming all cores active on a 8280 CPU (which is CascadeLake - same generation as used in c5 AWS instances). 

We can see that with all cores active, AVX code will run at 3.3GHz, AVX2 code will run at 2.9GHz, and AVX512 code will run at 2.4GHz. Even though, there is a frequency drop, we are operating on 2X the data at a time. So, if we spend enough time in the computation, then AVX512 will be a net benefit. In addition, AVX512 frequency is improving in subsequent generations - e.g. The m6i instance (IceLake ) is bridging the gap between AVX2 and AVX512 frequency. 

![AVX_frequency](https://user-images.githubusercontent.com/8259383/132895590-957b01f5-9cee-42ad-8bbe-bfb9b2bc590e.png)
![AVX2_frequency](https://user-images.githubusercontent.com/8259383/132894388-2500680c-3ab6-40be-a2bf-f2ad71cdd3d3.png)
![AVX512_frequency](https://user-images.githubusercontent.com/8259383/132894391-46e401b2-841c-4a81-87d3-8a8e3ac20ff2.png)

slice4e replied:
> Hm. It looks like this `#ifdef` does not have a corresponding `#endif` and this seems to trigger the CI build fail for some systems.

Thank you for catching this. I have added the missing `#endif`

slice4e replied:
The initial ann-bechmark results that I posted were with parallelism=3 on an c5.4xlarge instance. 
I just tested with parallelism=6 and the results are similar. I tried with parallelism=8, but we run out of memory on c5.4xlarge. 

![image](https://user-images.githubusercontent.com/8259383/132992341-38ac14ef-b014-497c-825f-bd7d3f8dfb9c.png)
![image](https://user-images.githubusercontent.com/8259383/132992350-df1885c8-9c29-4a54-805f-693875b34217.png)
![image](https://user-images.githubusercontent.com/8259383/132992356-c52d835a-fc40-4115-ac9c-76a0c24e7288.png)
![image](https://user-images.githubusercontent.com/8259383/132992359-442c0143-fc4a-456c-872e-511507ca9847.png)

yurymalkov replied:
Thanks for the test and explanation!

yurymalkov commented on hnswlib/space_l2.h:
```diff
@@ -19,7 +19,38 @@ namespace hnswlib {
+            v2 = _mm512_loadu_ps(pVect2);
+            pVect2 += 16;
+            diff = _mm512_sub_ps(v1, v2);
+            // sum = _mm512_fmadd_ps(diff, diff, sum);
```
I wonder why the line is commented? Does it perform slower?

slice4e replied:
Yes - I tested a fused multiply-add and it was slower for the AVX2 code. Was the same for the AVX512 code, so I decided to not use it.

yurymalkov reviewed (changes requested):
Thanks you so much for the PR! 
It seems like there is #endif missing. Can you please check?
Also, I wonder if you've tested it with sustained multithreaded load? I heard AVX512 can cause cloud instances to overheat and throttle, so they can even perform worse.

yurymalkov commented:
I've tried to run the code and noticed that `__AVX512__` is not defined in gcc with `-march=native` (there are `__AVX512BW__`, `__AVX512F__`, `__AVX512VNNI__` and many others, probably corresponding to different subsets). 
I wonder, how do you test the code? Can you please add instructions on how to compile with AVX512 support?

slice4e commented:
> I've tried to run the code and noticed that `__AVX512__` is not defined in gcc with `-march=native` (there are `__AVX512BW__`, `__AVX512F__`, `__AVX512VNNI__` and many others, probably corresponding to different subsets).
> I wonder, how do you test the code? Can you please add instructions on how to compile with AVX512 support?

Hi - Sorry, this was a mistake on my part. For some reason, I did not realize that we are checking for the presence of AVX/AVX2/etc at compile time, and I assumed that the application using hnswlib is setting the defines. In highsight, that does not make sense...

I fixed it now. We are now checking for __AVX512F__ (which is the foundational set of AVX512 instructions.). All the instructions that we are using are part of __AVX512F__
For my performance testing, I had simply hardcoded the #define USE_AVX512 ... in the AVX512 tests. 


In addition, I fixed a bug which was causing a crash on some system due to not aligning the memory correctly for the store instruction (_mm512_store_ps). This instruction requires memory to be aligned to 64 bytes.

yurymalkov commented on hnswlib/space_l2.h:
```diff
@@ -19,7 +19,38 @@ namespace hnswlib {
+
+    // Favor using AVX512 if available.
+    static float
+    L2SqrSIMD16Ext(float *pVect1, float *pVect2, size_t qty) {
```
Sorry for a long response. It seems `float` type here obstructs the compilation.
It should be `void *`

slice4e replied:
> Sorry for a long response. It seems `float` type here obstructs the compilation. It should be `void *`

Thanks for catching this. I have fixed it.

yurymalkov replied:
I've tried and there other errors with `constant  void *` to `void *` conversion (I guess it should be made const void * down the stack).
Can you please fix it and test it with `pip install .` and `python -m unittest discover --start-directory python_bindings/tests --pattern "*_test*.py"` (to run the tests) on a linux system.

slice4e replied:
Thanks for the help! I fixed those errors. I was able to compile it and test it without errors following your instructions.

yurymalkov commented:
Thanks for the PR!!
